cli/cmd/validate/validate.go

```diff
@@ -0,0 +1,385 @@
+	if err != nil {
+		return err
+	}
+	_, err = repo.ListGlob(ctx, "**", false)
+	if err != nil {
+		if errors.Is(err, drivers.ErrRepoListLimitExceeded) {
+			ch.PrintfError("The project directory exceeds the limit of %d files. Please open Rill against a directory with fewer files or set \"ignore_paths\" in rill.yaml.\n", drivers.RepoListLimit)
+		}
+		return fmt.Errorf("failed to list project files: %w", err)
+	}
+	return nil
+}
+
+func parseVariables(vals []string) (map[string]string, error) {
+	res := make(map[string]string)
+	for _, val := range vals {
+		parsed, err := godotenv.Unmarshal(val)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse variable %q: %w", val, err)
+		}
+		for k, v := range parsed {
+			res[k] = v
+		}
+	}
+	return res, nil
+}
+
+func parseProject(ctx context.Context, projectPath, environment string) ([]ParseError, error) {
+	repo, instanceID, err := cmdutil.RepoForProjectPath(projectPath)
+	if err != nil {
+		return nil, err
+	}
+
+	p, err := parser.Parse(ctx, repo, instanceID, environment, defaultOLAPConnector)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse project: %w", err)
+	}
+	if p.RillYAML == nil {
+		return nil, fmt.Errorf("failed to parse project: %w", parser.ErrRillYAMLNotFound)
+	}
+	if len(p.Errors) == 0 {
+		return nil, nil
+	}
+
+	var parseErrors []ParseError
+	for _, e := range p.Errors {
+		parseErr := ParseError{
+			Message:  e.Message,
+			FilePath: e.FilePath,
+			External: e.External,
+		}
+		// Convert start location if available
+		if e.StartLocation != nil {
+			parseErr.StartLocation = &CharLocation{
+				Line: e.StartLocation.Line,
+			}
+		}
+		parseErrors = append(parseErrors, parseErr)
+	}
+	return parseErrors, nil
+}
+
+func reconcileAndReport(ctx context.Context, ch *cmdutil.Helper, app *local.App, reconcileTimeout time.Duration, outputFormat, outputFile string) error {
+	ctrl, err := app.Runtime.Controller(ctx, app.Instance.ID)
+	if err != nil {
+		return err
+	}
+
+	// Create a context with timeout for reconciliation
+	reconcileCtx, cancel := context.WithTimeout(ctx, reconcileTimeout)
+	defer cancel()
+
+	timedOut := false
+	// Kick off reconciliation and wait for completion
+	if err := ctrl.Reconcile(reconcileCtx, runtime.GlobalProjectParserName); err != nil {
+		if errors.Is(err, context.DeadlineExceeded) {
+			timedOut = true
+		} else {
+			return fmt.Errorf("failed to start reconciliation: %w", err)
+		}
+	}
+
+	if !timedOut {
+		if err := ctrl.WaitUntilIdle(reconcileCtx, true); err != nil {
+			if errors.Is(err, context.DeadlineExceeded) {
+				timedOut = true
+			} else {
+				return fmt.Errorf("failed while waiting for reconciliation to finish: %w", err)
+			}
+		}
+	}
+
+	// List resources (use parent ctx with a timeout to avoid blocking indefinitely)
+	listCtx, listCancel := context.WithTimeout(ctx, listResourcesTimeout)
+	defer listCancel()
+
+	resources, err := ctrl.List(listCtx, "", "", false)
```

Comment:
Aren't the listings always really fast? If you see it hanging, probably there's a deeper issue we need to fix.